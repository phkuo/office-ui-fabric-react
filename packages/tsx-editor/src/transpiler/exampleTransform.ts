/**
 * classNamePattern - pattern to find the name of the class that should render
 * constNamePattern - pattern to find the name of the const that should render
 * identifierPattern - pattern to get all identifiers from the imports
 * importPattern - pattern to get all imports even if they have multilines
 */
export const classNamePattern = new RegExp('(?![var ])(.*)(?= = \\/\\*\\* @class \\*\\/ \\(function \\(_super\\))', 'g');
export const constNamePattern = new RegExp('(?![export var ])(.*)(?= = function)', 'g');
export const identifierPattern = new RegExp('(?![import {])([\\s\\S]*)(?=})');
export const importPattern = new RegExp("(?:import)([\\s\\S]+?)';", 'g');

export interface ITransformedExample {
  output?: string;
  error?: string;
}

interface IExampleData {
  [key: string]: string;
  exampleData: string;
  TestImages: string;
  PeopleExampleData: string;
}

const exampleDataFiles: IExampleData = {
  exampleData: require('!raw-loader!office-ui-fabric-react/lib/utilities/exampleData'),
  TestImages: require('!raw-loader!office-ui-fabric-react/lib/common/TestImages'),
  PeopleExampleData: require('!raw-loader!office-ui-fabric-react/lib/components/ExtendedPicker/examples/PeopleExampleData')
};

export function transformExample(example: string, id: string) {
  // RegExp need to reset because of global flag
  classNamePattern.lastIndex = 0;
  constNamePattern.lastIndex = 0;
  identifierPattern.lastIndex = 0;
  importPattern.lastIndex = 0;

  const identifiers: string[] = [];
  const imports: string[] = [];
  let temp;
  let className;
  const output: ITransformedExample = { output: undefined, error: undefined };

  example = example.replace("import * as React from 'react';", '');

  /**
   * Getting class name that should render, it will find all class names but will only
   * use the last one since the examples have the class that should render at the end.
   * If there is no such class then it will check for the last const which should be
   * the const that should be the one to render.
   */
  while ((temp = classNamePattern.exec(example))) {
    className = temp[0];
  }
  if (className === undefined) {
    while ((temp = constNamePattern.exec(example))) {
      className = temp[0];
    }
    if (className !== undefined) {
      className = className.replace('var ', '');
    }
  }

  /**
   * Getting all the imports from fabric and pushing them into an array.
   * If there are imports that do not come from OUFR then it will add an
   * error since the import is not supported.
   */
  while ((temp = importPattern.exec(example))) {
    let foundExampleImport = false;
    for (const filename of Object.keys(exampleDataFiles)) {
      if (temp[0].indexOf('/' + filename) !== -1) {
        example += `\n${exampleDataFiles[filename]}`;
        example = example.replace(temp[0], '');
        foundExampleImport = true;
      }
    }
    if (!/office-ui-fabric-react/.test(temp[0]) && !foundExampleImport) {
      output.error = `Error while transforming example: Unsupported import - ${temp[0]}.`;
    } else if (!foundExampleImport) {
      imports.push(temp[0]);
    }
  }

  /**
   * This will loop through all imports getting and saving the identifiers,
   * if no identifiers were found then it will just eliminate the import.
   * If identifiers were found, the string will be split, spaces will be removed,
   * and next line characters will also be removed.
   */
  imports.forEach((imp: string) => {
    temp = identifierPattern.exec(imp);
    if (temp !== null) {
      temp[0].split(',').forEach((ident: string) => {
        identifiers.push(ident.replace(/\s/g, ''));
      });
    }
    example = example.replace(imp, '');
  });

  // If there are exports in the code then they will be removed since they are not supported.
  example = example.replace(/^export /gm, '');

  /**
   * adding line to render React and adding identifiers.
   */
  example = `const { ${identifiers.join(', ')}, Fabric } = window.Fabric;\n
    ${example}
    ReactDOM.render(
      React.createElement(Fabric, null, React.createElement(${className}, null)),
      document.getElementById('${id}')
    );
    `;
  output.output = example;
  return output;
}
