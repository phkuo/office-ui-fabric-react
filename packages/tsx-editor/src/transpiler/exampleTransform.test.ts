import { transformExample, identifierPattern, importPattern, classNamePattern, constNamePattern } from './exampleTransform';
import * as fs from 'fs';
import * as path from 'path';

describe('example transform', () => {
  function transformFile(file: string) {
    const filename = path.resolve(__dirname, './examples/' + file);
    const fileContents = fs.readFileSync(filename).toString();
    return transformExample(fileContents, 'fake');
  }

  it('handles examples with function components', () => {
    const result = transformFile('function.txt');
    expect(result.output).toMatchSnapshot();
  });

  it('handles examples with class components', () => {
    const result = transformFile('class.txt');
    expect(result.output).toMatchSnapshot();
  });

  it('returns an error from relative imports', () => {
    const result = transformFile('relativeImport.txt');
    expect(result.error).toBe("Error while transforming example: Unsupported import - import '../../fake.scss';.");
  });

  it('handles transpiled examples with function components', () => {
    const result = transformFile('functionTranspiled.txt');
    expect(result.output).toMatchSnapshot();
  });

  it('handles transpiled examples with class components', () => {
    const result = transformFile('classTranspiled.txt');
    expect(result.output).toMatchSnapshot();
  });

  it('returns an error from relative imports in code that was transpiled', () => {
    const result = transformFile('realtiveImportTranspiled.txt');
    expect(result.error).toBe("Error while transforming example: Unsupported import - import '../../fake.scss';.");
  });

  it('detects class names', () => {
    const classNamePatternCopy = new RegExp(classNamePattern.source);
    // This is needed to get rid of the global flag from the RegExp
    expect(`var SpinButtonBasicExample = /** @class */ (function (_super) {
      __extends(SpinButtonBasicExample, _super);
      function SpinButtonBasicExample() {
          return _super !== null && _super.apply(this, arguments) || this;
      }
      SpinButtonBasicExample.prototype.render = function () {
          return (React.createElement("div", { style: { width: '400px' } },
              React.createElement(SpinButton, { defaultValue: "0",
              label: 'Basic SpinButton:', min: 0, max: 100, step: 1,
               iconProps: { iconName: 'IncreaseIndentLegacy' },
               incrementButtonAriaLabel: 'Increase value by 1',
               decrementButtonAriaLabel: 'Decrease value by 1' })));
      };
      return SpinButtonBasicExample;
      }(React.Component));
      export { SpinButtonBasicExample };`).toMatch(classNamePatternCopy);
    expect(`var ButtonDefaultExample = /** @class */ (function (_super) {
      __extends(ButtonDefaultExample, _super);
      function ButtonDefaultExample() {
          return _super !== null && _super.apply(this, arguments) || this;
      }
      ButtonDefaultExample.prototype.render = function () {
          var _a = this.props, disabled = _a.disabled,
          checked = _a.checked;
          return (React.createElement("div", { className: css(classNames.twoup) },
              React.createElement("div", null,
                  React.createElement(Label, null, "Standard"),
                  React.createElement(DefaultButton,
                    { "data-automation-id": "test", allowDisabledFocus: true,
                    disabled: disabled, checked: checked,
                    text: "Standard Button", onClick: this._alertClicked })),
              React.createElement("div", null,
                  React.createElement(Label, null, "Primary"),
                  React.createElement(PrimaryButton, {
                    "data-automation-id": "test", disabled: disabled,
                    checked: checked, text: "Primary Button",
                    onClick: this._alertClicked, allowDisabledFocus: true }))));
      };
      ButtonDefaultExample.prototype._alertClicked = function () {
          alert('Clicked');
      };
      return ButtonDefaultExample;
      }(React.Component));
      { ButtonDefaultExample };
      return React.createElement("div");`).toMatch(classNamePatternCopy);
    expect(`import * as React from 'react';
      import { Label } from 'office-ui-fabric-react/lib/Label';
      export var LabelBasicExample = function () {
      return React.createElement("div");
    };
    `).not.toMatch(classNamePatternCopy);
  });

  it('detects const names', () => {
    const constNamePatternCopy = new RegExp(constNamePattern.source);
    expect(`import * as React from 'react';
      import { Label } from 'office-ui-fabric-react/lib/Label';
      export var LabelBasicExample = function () {
          return React.createElement("div");
        };
    `).toMatch(constNamePatternCopy);
    expect(`var stackTokens = { childrenGap: 20 };
      var DropdownBasicExample = function () {
      return (React.createElement(Stack, { tokens: stackTokens },
      React.createElement(Dropdown, { placeholder: "Select an option",
      label: "Basic uncontrolled example",
      options: options, styles: dropdownStyles }),
      React.createElement(Dropdown, {
        label: "Disabled example with defaultSelectedKey",
        defaultSelectedKey: "broccoli", options: options,
        disabled: true, styles: dropdownStyles }),
      React.createElement(Dropdown, {
        placeholder: "Select options",
        label: "Multi-select uncontrolled example",
        defaultSelectedKeys: ['apple', 'banana', 'grape'],
        multiSelect: true, options: options,
        styles: dropdownStyles })));
      };
    return React.createElement("div");`).toMatch(constNamePatternCopy);
    expect(`var SpinButtonBasicExample = /** @class */ (function (_super) {
      __extends(SpinButtonBasicExample, _super);
      function SpinButtonBasicExample() {
          return _super !== null && _super.apply(this, arguments) || this;
      }
      return SpinButtonBasicExample;
      }(React.Component));
      export { SpinButtonBasicExample };`).not.toMatch(constNamePatternCopy);
  });

  it('detects imports', () => {
    const importPatternCopy = new RegExp(importPattern.source);
    expect(`import { exampleData } from '../../exampleData';`).toMatch(importPatternCopy);
    expect(`import { Label, Stack, Button } from 'office-ui-fabric-react';`).toMatch(importPatternCopy);
    expect(`const x = require('../../fake.scss');`).not.toMatch(importPatternCopy);
  });

  it('detects identifiers', () => {
    const identifierPatternCopy = new RegExp(identifierPattern.source);
    expect(`import { Label, Stack, Button } from 'office-ui-fabric-react';`).toMatch(identifierPatternCopy);
    expect(`import { exampleData } from '../../exampleData';`).toMatch(identifierPatternCopy);
    expect(`import '../../fake';`).not.toMatch(identifierPatternCopy);
  });
});
