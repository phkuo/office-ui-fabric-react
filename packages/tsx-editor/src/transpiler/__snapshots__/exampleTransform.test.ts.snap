// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example transform handles examples with class components 1`] = `
"const { SpinButton, Fabric } = window.Fabric;

    


class SpinButtonBasicExample extends React.Component<any, any> {
  public render(): JSX.Element {
    return (
      <div style={{ width: '400px' }}>
        <SpinButton
          defaultValue=\\"0\\"
          label={'Basic SpinButton:'}
          min={0}
          max={100}
          step={1}
          iconProps={{ iconName: 'IncreaseIndentLegacy' }}
          incrementButtonAriaLabel={'Increase value by 1'}
          decrementButtonAriaLabel={'Decrease value by 1'}
        />
      </div>
    );
  }
}

    ReactDOM.render(
      React.createElement(Fabric, null, React.createElement(undefined, null)),
      document.getElementById('fake')
    );
    "
`;

exports[`example transform handles examples with function components 1`] = `
"const { Label, Fabric } = window.Fabric;

    


const LabelBasicExample = () => {
  return (
    <div>
      <Label>I'm a Label</Label>
    </div>
  );
};

    ReactDOM.render(
      React.createElement(Fabric, null, React.createElement(undefined, null)),
      document.getElementById('fake')
    );
    "
`;

exports[`example transform handles transpiled examples with class components 1`] = `
"const { SpinButton, Fabric } = window.Fabric;

    var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();


var SpinButtonBasicExample = /** @class */ (function (_super) {
    __extends(SpinButtonBasicExample, _super);
    function SpinButtonBasicExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SpinButtonBasicExample.prototype.render = function () {
        return (React.createElement(\\"div\\", { style: { width: '400px' } },
            React.createElement(SpinButton, { defaultValue: \\"0\\", label: 'Basic SpinButton:', min: 0, max: 100, step: 1, iconProps: { iconName: 'IncreaseIndentLegacy' }, incrementButtonAriaLabel: 'Increase value by 1', decrementButtonAriaLabel: 'Decrease value by 1' })));
    };
    return SpinButtonBasicExample;
}(React.Component));
{ SpinButtonBasicExample };

    ReactDOM.render(
      React.createElement(Fabric, null, React.createElement(SpinButtonBasicExample, null)),
      document.getElementById('fake')
    );
    "
`;

exports[`example transform handles transpiled examples with function components 1`] = `
"const { Label, Fabric } = window.Fabric;

    

var LabelBasicExample = function () {
    return (React.createElement(\\"div\\", null,
        React.createElement(Label, null, \\"I'm a Label\\")));
};

    ReactDOM.render(
      React.createElement(Fabric, null, React.createElement(LabelBasicExample, null)),
      document.getElementById('fake')
    );
    "
`;
